dev/dram/device.c

```diff
@@ -93,75 +112,429 @@ void dram_cbt_exit(dram_dev_t *dram,
 {
-    *offset = (dram_cfg->phy_rd_en + CA_LAST);
+    *offset = (dram_cfg->phy_rd_en + CA_CLK_LEN);
 }
 
 void dram_get_write_enable_offset_timing(dram_freq_cfg_t *dram_cfg,
                                          uint8_t *offset)
 {
-    *offset = (dram_cfg->phy_wr_lat + CA_LAST);
+    *offset = (dram_cfg->phy_wr_lat + CA_CLK_LEN);
 }
 
 void dram_get_write_data_offset_timing(dram_freq_cfg_t *dram_cfg,
                                        uint8_t *offset)
 {
-    *offset = (dram_cfg->phy_wr_lat + dram_cfg->phy_wr_en + CA_LAST);
+    *offset = (dram_cfg->phy_wr_lat + dram_cfg->phy_wr_en + CA_CLK_LEN);
 }
 
-static void dram_write_mode_register(__UNUSED__ dram_dev_t *dram,
-                                 dfi_buffer_dev_t *dfi_buffer,
-                                 dram_freq_cfg_t *cfg,
-                                 uint8_t mr,
-                                 uint8_t op)
+wddr_return_t dram_prepare_cbt_sequence(dram_dev_t *dram,
+                               dfi_tx_packet_buffer_t *buffer,
+                               chipselect_t cs,
+                               uint8_t vref_setting,
+                               uint8_t command_address)
+{
+    uint8_t ratio = dram->cfg->ratio;
+    packet_item_t *packet;
+    uint8_t offset_remainder;
+    uint8_t time_offset = 1;
+    uint16_t initial_dram_offset = 0;
+    uint16_t dram_clk = 0;
+    packet_group_info_t wrdata_en_info, wrdata_info;
+    uint8_t cycles_per_packet = 1 << ratio;
+    uint8_t data_src[cycles_per_packet];
+    command_data_t data = {0};
+    memset(&data_src[0], vref_setting, cycles_per_packet);
+    create_data_frame(&data, WDDR_DQ_BYTE_0, &data_src[0], cycles_per_packet, 0);
+
+    // Calculate timestamp of first CA packet (time 0).
+    // Shifted by ratio to be measured in DRAM CLK cycles.
+    initial_dram_offset = (buffer->ts_last_packet + time_offset) << ratio;
+
+    create_ck_packet_sequence(buffer, time_offset);
+
+    dram_clk = initial_dram_offset + cycles_per_packet;
+
+    create_packet_group_info(&wrdata_info, ratio, dram_clk, cycles_per_packet);
+
+    // Generate DQ data packets for setting new CA VREF value
+    // First Packet is setting DQ pins to CA VREF value
+    packet = create_ck_packet_sequence(buffer, 1);
+    if (packet == NULL)
+    {
+        return WDDR_ERROR;
+    }
+
+    fill_wrdata_packet(packet, &wrdata_info, cs, &data, cycles_per_packet, cycles_per_packet);
+
+    // Second packet is toggling DQS tDStrain after setting DQ pins
+    // Need to extend offset by 2 to account for preamble
+    dram_clk += dram->cfg->t_sh_train;
+    create_packet_group_info(&wrdata_en_info, ratio, dram_clk, 4);
+    create_packet_group_info(&wrdata_info, ratio, dram_clk, PACKET_MAX_NUM_PHASES);
+
+    if (wrdata_en_info.cycle_offset > 0)
+    {
+        wrdata_en_info.dfi_ts_start++;
+        wrdata_en_info.cycle_offset = 0;
+        wrdata_info.dfi_ts_start++;
+        wrdata_info.cycle_offset = 0;
+    }
```

Comment:
Is this still needed? I think we're handling this now that we're using masks and such.

Reply:
We can have a discussion on this if needed.  I left this here to make the packet generation cleaner, where I don't have to account for filling wrdata extra at the end.  It keeps the generation and offsets somewhat easier if everything starts at phase 0 (this is moreso applicable when 1TO4 comes into play).  We can remove it and ensure wrdata is long enough to fill packets such that there is no gap afterwards, but timing is easier when not having to understand offsets from previous packet or anything like that.

Reply:
I'm not sure I follow. Let's discuss it.

Reply:
This has been removed and fixed with the new do..while loop encompasing entire length of DQ/DQS packets.